rfcs/0045-zkapp-balance-data-in-archive.md

```diff
@@ -0,0 +1,292 @@
+[detailed-design]: #detailed-design
+
+### Changes to the archive database schema
+
+Balances are no longer per-transaction, but per-block.
+
+Table `blocks_internal_commands`, remove the column:
+
+  `receiver_account_creation_fee_paid`
+  `receiver_balance`
+
+Table `blocks_user_commands`, remove the columns:
+
+  `fee_payer_account_creation_fee_paid`
+  `receiver_account_creation_fee_paid`
+  `fee_payer_balance`
+  `source_balance`
+  `receiver_balance`
+
+where the array contains account creation fees for each of the `other_parties`.
+While the array is not nullable, array elements may be `NULL`.
+
+Rename the existing `zkapp_account` to `zkapp_precondition_accounts`, and in the
+table `zkapp_predicate`, rename `account_id` to `precondition_account_id`, and
+modify the foreign key reference accordingly.
+
+Add new table `zkapp_accounts`:
+```
+  app_state           int     NOT NULL  REFERENCES zkapp_states
+  verification_key    int     NOT NULL  REFERENCES zkapp_verification_keys(id)
+  zkapp_version       bigint  NOT NULL
+  sequence_state      int     NOT NULL  REFERENCES zkapp_sequence_states(id)
+  last_sequence_slot  bigint  NOT NULL
+  proved_state        bool    NOT NULL
+```
+
+The table `balances` is replaced by a new table `accounts_accessed`, with columns:
+```
+  block_id            int                NOT NULL  REFERENCES blocks(id)
+  public_key          int                NOT NULL  REFERENCES public_keys(id)
+  token               int                NOT NULL  REFERENCES tokens(id)
+  token_permissions   token_permissions  NOT NULL
+  balance             bigint             NOT NULL
+  nonce               bigint             NOT NULL
+  receipt_chain_hash  text               NOT NULL
+  delegate            int                          REFERENCES public_keys(id)
+  voting_for          text    NOT NULL
+  timing              int                          REFERENCES timing_info(id)
+  permissions         int     NOT NULL             REFERENCES zkapp_permissions(id)
+  zkapp               int                          REFERENCES zkapp_accounts(id)
+  zkapp_uri           text    NOT NULL
+```
+
+The new table `zkapp_sequence_states` has the same definition as the existing `zkapp_states`;
+it represents a vector of field elements.  We probably don't want to commingle sequence states
+with app states in a single table, because they contain differing numbers of elements.
+
+The new type `token_permissions` is:
+
+  `CREATE TYPE token_permission AS ENUM ('token_owned_new_accounts_disabled', 'token_owned_new_accounts_enabled',
+                                         'token_not_owned_account_disabled', 'token_not_owned_account_enabled')`
+
+(Alternatively, we could create a new table with four entries corresponding to these possibilities,
+and refer to it with a foreign key.)
+
+The new table `tokens` is:
+```
+  id                  serial  PRIMARY KEY
+  token_id            text    NOT NULL
+  token_symbol        text    NOT NULL
```

Comment:
When this changes, this change should propagate to each account in that token. Instead of doing this -- which intuitively seems pretty expensive -- it may make sense to point to the 'owner account' and retrieve the latest symbol from there so that it's always up-to-date.

Reply:
How about this change to `accounts_accessed`:
```
  id                      serial  PRIMARY KEY 
...
  token_owner_account_id  int               REFERENCES accounts_accessed(id)                                                                                                                                                                  
  token_id                text    NOT NULL                                                                                                                                                                                                    
  token_symbol            text    NOT NULL
...                                    
```
Invariant: `token_owner_account` is `NULL` iff `token_id` and `token_symbol` are not `NULL`.
That is, the token owner account has values for the token id and symbol, while non-owner accounts
refer to the token owner account.